// google-ads-api/lib/tests/campaignNegatives.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = __importDefault(require(".."));
const config_1 = __importDefault(require("../config"));
jest.setTimeout(30000);
const getRandomKeywordText = () => `test-negative-${(Math.random() * 1000000 + 1).toFixed(0)}`;
// const log = (obj: object) => {
//     console.log(require('util').inspect(obj, false, null))
// }
describe('Campaign Negatives', () => __awaiter(this, void 0, void 0, function* () {
    const lib_instance = new __1.default({
        client_id: config_1.default.client_id,
        client_secret: config_1.default.client_secret,
        developer_token: config_1.default.developer_token,
    });
    const customer = lib_instance.Customer({
        customer_account_id: config_1.default.cid,
        manager_cid: config_1.default.manager_cid,
        refresh_token: config_1.default.refresh_token,
    });
    const campaign_id = 1485014801;
    let criterion_id = '';
    it('Lists Campaign Negatives', () => __awaiter(this, void 0, void 0, function* () {
        expect.assertions(1);
        const criterions = yield customer.campaignNegatives.list({
            constraints: ['campaign.id =' + campaign_id, { type: 'KEYWORD' }],
        });
        expect(criterions).toBeInstanceOf(Array);
    }));
    it('Creates New Campaign Negative', (done) => __awaiter(this, void 0, void 0, function* () {
        expect.assertions(1);
        const new_criterion = yield customer.campaignNegatives.create({
            campaign_id,
            negative: true,
            keyword: {
                text: getRandomKeywordText(),
                match_type: 'EXACT',
            },
        });
        expect(new_criterion).toEqual({
            id: expect.any(String),
            resource_name: expect.any(String),
        });
        criterion_id = new_criterion.id;
        done();
    }));
    it('Creates Multiple New Campaign Negatives', (done) => __awaiter(this, void 0, void 0, function* () {
        expect.assertions(1);
        const new_criteria = yield customer.campaignNegatives.create([
            {
                campaign_id,
                negative: true,
                keyword: {
                    text: getRandomKeywordText(),
                    match_type: 'EXACT',
                },
            },
            {
                campaign_id,
                negative: true,
                keyword: {
                    text: getRandomKeywordText(),
                    match_type: 'EXACT',
                },
            },
        ]);
        expect(new_criteria).toContainEqual(expect.objectContaining({
            id: expect.any(String),
            resource_name: expect.any(String),
        }));
        done();
    }));
    it('Retrieves Single Campaign Negative', (done) => __awaiter(this, void 0, void 0, function* () {
        expect.assertions(1);
        const negative_keyword = yield customer.campaignNegatives.retrieve(criterion_id);
        expect(criterion_id).toContain(negative_keyword.criterion_id);
        done();
    }));
}));
