// google-ads-api/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADWORDS_API_BASE_URL = 'https://googleads.googleapis.com/v0/customers/';
exports.ADWORDS_AUTH_URL = 'https://accounts.google.com/o/oauth2/token';
var ENDPOINTS;
(function (ENDPOINTS) {
    ENDPOINTS["search"] = "googleAds:search";
    ENDPOINTS["campaigns"] = "campaigns";
    ENDPOINTS["campaign_budgets"] = "campaignBudgets";
    ENDPOINTS["campaign_criteria"] = "campaignCriteria";
    ENDPOINTS["campaign_shared_sets"] = "campaignSharedSets";
    ENDPOINTS["customers"] = "customers";
    ENDPOINTS["adgroups"] = "adGroups";
    ENDPOINTS["ads"] = "adGroupAds";
    ENDPOINTS["adgroup_critera"] = "adGroupCriteria";
    ENDPOINTS["shared_sets"] = "sharedSets";
    ENDPOINTS["shared_set_criteria"] = "sharedCriteria";
})(ENDPOINTS = exports.ENDPOINTS || (exports.ENDPOINTS = {}));
var RESOURCE_NAMES;
(function (RESOURCE_NAMES) {
    RESOURCE_NAMES["campaign"] = "campaign";
    RESOURCE_NAMES["campaign_budget"] = "campaign_budget";
    RESOURCE_NAMES["campaign_criteria"] = "campaign_criterion";
    RESOURCE_NAMES["campaign_shared_set"] = "campaign_shared_set";
    RESOURCE_NAMES["adgroup"] = "ad_group";
    RESOURCE_NAMES["ad"] = "ad_group_ad";
    RESOURCE_NAMES["adgroup_critera"] = "ad_group_criterion";
    RESOURCE_NAMES["shared_set"] = "shared_set";
    RESOURCE_NAMES["shared_set_criterion"] = "shared_criterion";
})(RESOURCE_NAMES = exports.RESOURCE_NAMES || (exports.RESOURCE_NAMES = {}));
