// google-ads-api/lib/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const all_metrics = [
    { name: 'clicks', is_number: true },
    { name: 'conversions', is_number: true },
    { name: 'all_conversions', is_number: true },
    { name: 'cost_micros', is_number: true },
    { name: 'impressions', is_number: true },
    {
        name: 'cost',
        is_custom: true,
        is_micros: true,
        is_number: true,
        pre_query_hook: (report) => {
            if (report.metrics) {
                report.metrics.push('metrics.cost_micros');
            }
            report.metrics = lodash_1.reject(report.metrics, i => i === 'metrics.cost');
            if (!report.constraints || lodash_1.isString(report.constraints)) {
                return report;
            }
            const cost_constraint = lodash_1.find(report.constraints, {
                key: 'metrics.cost',
            });
            if (!cost_constraint) {
                return report;
            }
            report.constraints.push({
                key: 'metrics.cost_micros',
                op: cost_constraint.op,
                val: +cost_constraint.val * 1000000,
            });
            report.constraints = lodash_1.reject(report.constraints, { key: 'metrics.cost' });
            return report;
        },
    },
    { name: 'average_cpc', is_micros: true, is_number: true },
];
exports.default = all_metrics;
