// google-ads-api/lib/entities/CampaignSharedSets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CampaignSharedSets {
    /**
     * CampaignSharedSets
     * @constructor
     */
    constructor(http_controller, endpoint, resource) {
        this.http_controller = http_controller;
        this.endpoint = endpoint;
        this.resource = resource;
    }
    /**
     * Links Shared Set with Campaign
     * @param {object} config
     * @param {number|string} config.campaign_id - ID of the campaign
     * @param {number|string} config.shared_set_id - ID of the shared set
     */
    link(config) {
        return this.http_controller.create(config, this.endpoint);
    }
    /**
     * Lists all Campaign Shared Sets
     * @param {object} config
     * @param {integer} config.limit
     * @param {object} config.constraints
     */
    list(config) {
        return this.http_controller.list(config, this.resource);
    }
    /**
     * Retrieves Single Campaign Shared Set
     * @param {string} campaign_shared_set_id
     */
    retrieve(campaign_shared_set_id) {
        return this.http_controller.retrieve(this.endpoint, campaign_shared_set_id);
    }
    /**
     * Unlinks Shared Set from Campaign
     * @param {string} campaign_shared_set_id
     */
    delete(campaign_shared_set_id) {
        return this.http_controller.delete(this.endpoint, campaign_shared_set_id);
    }
}
exports.default = CampaignSharedSets;
